File: src/main/java/io/veridia/jsonlogic/JsonLogic.java
package io.veridia.jsonlogic;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import io.veridia.jsonlogic.helpers.ToBoolean;

/**
 * Entry point for evaluating JSON Logic expressions.
 * <p>
 * Instances manage parsing, compilation, and a small cache of compiled expressions.
 */
public class JsonLogic {
    private final transient ObjectMapper mapper = new ObjectMapper();
    private final transient LogicParser parser = new LogicParser();
    private final transient JsonLogicCompiler compiler = new JsonLogicCompiler();

    private final transient Cache<String, CompiledExpression> compiledExpressionsCache;

    /**
     * Creates a new {@code JsonLogic} instance with a default compiled expression cache.
     */
    public JsonLogic() {
        this.compiledExpressionsCache = CacheBuilder.newBuilder()
                .maximumSize(1000)
                .build();
    }

    /**
     * Applies a JSON Logic expression against the provided context.
     *
     * @param str JSON Logic expression as a string
     * @param ctx context object or JSON string to use when evaluating the expression
     * @return the raw evaluation result, or {@code null} when the expression is blank
     * @throws JsonProcessingException when the expression or context cannot be parsed
     */
    public Object apply(String str, Object ctx) throws JsonProcessingException {
        if (str == null || str.isBlank())
            return null;

        CompiledExpression expr = compile(str);

        Object context = ctx == null ? null : ctx instanceof String ? mapper.readValue((String) ctx, new TypeReference<>() {}) : ctx;

        return expr.eval(context);
    }

    /**
     * Parses and compiles a JSON Logic expression, returning a reusable {@link CompiledExpression}.
     * <p>
     * Compilation results are cached, so repeated calls with the same expression string are cheap.
     * Callers on hot paths should compile once and invoke {@link CompiledExpression#eval(Object)}
     * directly against pre-parsed contexts to avoid all per-evaluation parsing and cache lookups.
     *
     * @param str JSON Logic expression as a string
     * @return the compiled expression (a no-op returning {@code null} for a blank input)
     * @throws JsonProcessingException when the expression cannot be parsed
     */
    public CompiledExpression compile(String str) throws JsonProcessingException {
        if (str == null || str.isBlank())
            return ctx -> null;

        // The expression string is the cache key directly: String.hashCode is cached by the JVM,
        // so lookups no longer pay for hashing the full expression on every call.
        CompiledExpression expr = compiledExpressionsCache.getIfPresent(str);

        if (expr == null) {
            JsonNode json = mapper.readTree(str);
            expr = compiler.compile(parser.parse(json));
            compiledExpressionsCache.put(str, expr);
        }

        return expr;
    }

    /**
     * Applies a JSON Logic expression and coerces the result to a boolean.
     *
     * @param str JSON Logic expression as a string
     * @param ctx context object or JSON string to use when evaluating the expression
     * @return boolean interpretation of the evaluation result
     * @throws JsonProcessingException when the expression or context cannot be parsed
     */
    public boolean check(String str, Object ctx) throws JsonProcessingException {
        Object result = apply(str, ctx);

        return ToBoolean.eval(result);
    }

    /**
     * Registers a custom operator implementation for use during compilation.
     *
     * @param impl operator implementation to register
     * @return this instance for chaining
     */
    public JsonLogic registerOperator(Operator impl) {
        compiler.registerOperator(impl);

        return this;
    }
}

File: src/main/scala/io/veridia/jsonlogic/JsonLogic.scala
package io.veridia.jsonlogic

import io.circe.parser.parse as parseJson
import io.veridia.jsonlogic.cache.BoundedCache
import io.veridia.jsonlogic.helpers.ToBoolean
import io.veridia.jsonlogic.values.Values

/** Entry point for evaluating JSON Logic expressions. Manages parsing, compilation, and a small
  * cache of compiled expressions — construction and every public method are safe to call
  * concurrently, and callers are expected to share one instance across many evaluations.
  */
final class JsonLogic:
  private val compiler = new JsonLogicCompiler
  private val cache = new BoundedCache[String, CompiledExpression](1000)

  /** Parses and compiles a JSON Logic expression, returning a reusable [[CompiledExpression]].
    * Results are cached per instance, keyed by the raw expression string; callers on hot paths
    * should compile once and invoke `eval` directly against pre-parsed contexts.
    */
  def compile(str: String): CompiledExpression =
    if str == null || str.isBlank then (_ => null)
    else
      val cached = cache.getIfPresent(str)
      if cached != null then cached
      else
        val json = parseJson(str).getOrElse(throw new JsonLogicException(s"Invalid JSON Logic expression: $str"))
        val compiled = compiler.compile(LogicParser.parse(json))
        cache.put(str, compiled)
        compiled

  /** Applies a JSON Logic expression against the provided context.
    *
    * @param str JSON Logic expression as a string
    * @param ctx context object (`Map`/`List`/POJO-free structure) or a JSON string to evaluate against
    * @return the raw evaluation result, or `null` when the expression is blank
    */
  def apply(str: String, ctx: Object): Object =
    if str == null || str.isBlank then null
    else
      val context: Object = ctx match
        case null => null
        case s: String =>
          parseJson(s).map(Values.jsonToValue).getOrElse(throw new JsonLogicException(s"Invalid JSON context: $s"))
        case other => other
      compile(str).eval(context)

  /** Applies a JSON Logic expression and coerces the result to a boolean. */
  def check(str: String, ctx: Object): Boolean = ToBoolean.eval(apply(str, ctx))

  /** Registers a custom operator implementation for use during compilation.
    *
    * @return this instance, for chaining
    */
  def registerOperator(impl: Operator): JsonLogic =
    compiler.registerOperator(impl)
    this
